File: tests/valid/TestUserStory.java
package com.example.tests;

import java.util.regex.Pattern;
import java.util.concurrent.TimeUnit;
import org.junit.*;
import static org.junit.Assert.*;
import static org.hamcrest.CoreMatchers.*;
import org.openqa.selenium.*;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.Select;

public class TestUserStory {
  private WebDriver driver;
  private String baseUrl;
  private boolean acceptNextAlert = true;
  private StringBuffer verificationErrors = new StringBuffer();

  @Before
  public void setUp() throws Exception {
    driver = new FirefoxDriver();
    baseUrl = "http://localhost/app/app/";
    driver.manage().timeouts().implicitlyWait(30, TimeUnit.SECONDS);
  }

  @Test
  public void testUserStory() throws Exception {
    driver.get(baseUrl + "");
    driver.findElement(By.linkText("Backlog")).click();
    try {
      assertTrue(driver.getCurrentUrl().matches("^http://localhost/app/app/[\\s\\S]userstories$"));
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertFalse(isElementPresent(By.cssSelector("table.table-list")));
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    driver.findElement(By.cssSelector("button.btn.btn-link")).click();
    try {
      assertTrue(driver.getCurrentUrl().matches("^http://localhost/app/app/[\\s\\S]userstories&add$"));
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    driver.findElement(By.name("new_us_identifier")).clear();
    driver.findElement(By.name("new_us_identifier")).sendKeys("us1");
    driver.findElement(By.name("new_us_summary")).clear();
    driver.findElement(By.name("new_us_summary")).sendKeys("je veux....");
    driver.findElement(By.name("new_us_priority")).clear();
    driver.findElement(By.name("new_us_priority")).sendKeys("1");
    driver.findElement(By.name("new_us_difficulty")).clear();
    driver.findElement(By.name("new_us_difficulty")).sendKeys("2");
    driver.findElement(By.name("new_us_description")).clear();
    driver.findElement(By.name("new_us_description")).sendKeys("descr");
    driver.findElement(By.cssSelector("button.btn.btn-lg")).click();
    try {
      assertTrue(driver.getCurrentUrl().matches("^http://localhost/app/app/[\\s\\S]userstories$"));
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertTrue(isElementPresent(By.cssSelector("table.table-list")));
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("us1", driver.findElement(By.cssSelector("td.table-td")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    driver.findElement(By.name("VIEW")).click();
    try {
      assertEquals("us1", driver.findElement(By.id("view_us_identifier")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("je veux....", driver.findElement(By.id("view_us_summary")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("1", driver.findElement(By.id("view_us_priority")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("2", driver.findElement(By.id("view_us_difficulty")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("todo", driver.findElement(By.id("view_us_state")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("descr", driver.findElement(By.id("view_us_description")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("", driver.findElement(By.id("view_us_duration")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    driver.get("http://localhost/app/app/?userstories");
    driver.findElement(By.cssSelector("a[name=\"EDIT\"] > img")).click();
    driver.findElement(By.name("edit_us_identifier")).clear();
    driver.findElement(By.name("edit_us_identifier")).sendKeys("us2");
    driver.findElement(By.name("edit_us_summary")).clear();
    driver.findElement(By.name("edit_us_summary")).sendKeys("je souhaite...");
    driver.findElement(By.name("edit_us_priority")).clear();
    driver.findElement(By.name("edit_us_priority")).sendKeys("2");
    driver.findElement(By.name("edit_us_difficulty")).clear();
    driver.findElement(By.name("edit_us_difficulty")).sendKeys("1");
    driver.findElement(By.name("edit_us_description")).clear();
    driver.findElement(By.name("edit_us_description")).sendKeys("descr2");
    driver.findElement(By.cssSelector("button.btn.btn-lg")).click();
    try {
      assertTrue(driver.getCurrentUrl().matches("^http://localhost/app/app/[\\s\\S]userstories$"));
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    driver.findElement(By.name("VIEW")).click();
    try {
      assertEquals("us2", driver.findElement(By.id("view_us_identifier")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("je souhaite...", driver.findElement(By.id("view_us_summary")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("2", driver.findElement(By.id("view_us_priority")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("1", driver.findElement(By.id("view_us_difficulty")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("todo", driver.findElement(By.id("view_us_state")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("descr2", driver.findElement(By.id("view_us_description")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    try {
      assertEquals("", driver.findElement(By.id("view_us_duration")).getText());
    } catch (Error e) {
      verificationErrors.append(e.toString());
    }
    driver.get("http://localhost/app/app/?userstories");
    driver.findElement(By.name("DEL")).click();
  }

  @After
  public void tearDown() throws Exception {
    driver.quit();
    String verificationErrorString = verificationErrors.toString();
    if (!"".equals(verificationErrorString)) {
      fail(verificationErrorString);
    }
  }

  private boolean isElementPresent(By by) {
    try {
      driver.findElement(by);
      return true;
    } catch (NoSuchElementException e) {
      return false;
    }
  }

  private boolean isAlertPresent() {
    try {
      driver.switchTo().alert();
      return true;
    } catch (NoAlertPresentException e) {
      return false;
    }
  }

  private String closeAlertAndGetItsText() {
    try {
      Alert alert = driver.switchTo().alert();
      String alertText = alert.getText();
      if (acceptNextAlert) {
        alert.accept();
      } else {
        alert.dismiss();
      }
      return alertText;
    } finally {
      acceptNextAlert = true;
    }
  }
}

File: tests/valid/TestUserStory.py
# -*- coding: utf-8 -*-
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import Select
from selenium.common.exceptions import NoSuchElementException
from selenium.common.exceptions import NoAlertPresentException
import unittest, time, re

class TestUserStory(unittest.TestCase):
    def setUp(self):
        self.driver = webdriver.Firefox()
        self.driver.implicitly_wait(30)
        self.base_url = "http://localhost/app/app/"
        self.verificationErrors = []
        self.accept_next_alert = True
    
    def test_user_story(self):
        driver = self.driver
        driver.get(self.base_url + "")
        driver.find_element_by_link_text("Backlog").click()
        try: self.assertRegexpMatches(driver.current_url, r"^http://localhost/app/app/[\s\S]userstories$")
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertFalse(self.is_element_present(By.CSS_SELECTOR, "table.table-list"))
        except AssertionError as e: self.verificationErrors.append(str(e))
        driver.find_element_by_css_selector("button.btn.btn-link").click()
        try: self.assertRegexpMatches(driver.current_url, r"^http://localhost/app/app/[\s\S]userstories&add$")
        except AssertionError as e: self.verificationErrors.append(str(e))
        driver.find_element_by_name("new_us_identifier").clear()
        driver.find_element_by_name("new_us_identifier").send_keys("us1")
        driver.find_element_by_name("new_us_summary").clear()
        driver.find_element_by_name("new_us_summary").send_keys("je veux....")
        driver.find_element_by_name("new_us_priority").clear()
        driver.find_element_by_name("new_us_priority").send_keys("1")
        driver.find_element_by_name("new_us_difficulty").clear()
        driver.find_element_by_name("new_us_difficulty").send_keys("2")
        driver.find_element_by_name("new_us_description").clear()
        driver.find_element_by_name("new_us_description").send_keys("descr")
        driver.find_element_by_css_selector("button.btn.btn-lg").click()
        try: self.assertRegexpMatches(driver.current_url, r"^http://localhost/app/app/[\s\S]userstories$")
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertTrue(self.is_element_present(By.CSS_SELECTOR, "table.table-list"))
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("us1", driver.find_element_by_css_selector("td.table-td").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        driver.find_element_by_name("VIEW").click()
        try: self.assertEqual("us1", driver.find_element_by_id("view_us_identifier").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("je veux....", driver.find_element_by_id("view_us_summary").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("1", driver.find_element_by_id("view_us_priority").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("2", driver.find_element_by_id("view_us_difficulty").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("todo", driver.find_element_by_id("view_us_state").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("descr", driver.find_element_by_id("view_us_description").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("", driver.find_element_by_id("view_us_duration").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        driver.get("http://localhost/app/app/?userstories")
        driver.find_element_by_css_selector("a[name=\"EDIT\"] > img").click()
        driver.find_element_by_name("edit_us_identifier").clear()
        driver.find_element_by_name("edit_us_identifier").send_keys("us2")
        driver.find_element_by_name("edit_us_summary").clear()
        driver.find_element_by_name("edit_us_summary").send_keys("je souhaite...")
        driver.find_element_by_name("edit_us_priority").clear()
        driver.find_element_by_name("edit_us_priority").send_keys("2")
        driver.find_element_by_name("edit_us_difficulty").clear()
        driver.find_element_by_name("edit_us_difficulty").send_keys("1")
        driver.find_element_by_name("edit_us_description").clear()
        driver.find_element_by_name("edit_us_description").send_keys("descr2")
        driver.find_element_by_css_selector("button.btn.btn-lg").click()
        try: self.assertRegexpMatches(driver.current_url, r"^http://localhost/app/app/[\s\S]userstories$")
        except AssertionError as e: self.verificationErrors.append(str(e))
        driver.find_element_by_name("VIEW").click()
        try: self.assertEqual("us2", driver.find_element_by_id("view_us_identifier").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("je souhaite...", driver.find_element_by_id("view_us_summary").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("2", driver.find_element_by_id("view_us_priority").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("1", driver.find_element_by_id("view_us_difficulty").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("todo", driver.find_element_by_id("view_us_state").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("descr2", driver.find_element_by_id("view_us_description").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        try: self.assertEqual("", driver.find_element_by_id("view_us_duration").text)
        except AssertionError as e: self.verificationErrors.append(str(e))
        driver.get("http://localhost/app/app/?userstories")
        driver.find_element_by_name("DEL").click()
    
    def is_element_present(self, how, what):
        try: self.driver.find_element(by=how, value=what)
        except NoSuchElementException, e: return False
        return True
    
    def is_alert_present(self):
        try: self.driver.switch_to_alert()
        except NoAlertPresentException, e: return False
        return True
    
    def close_alert_and_get_its_text(self):
        try:
            alert = self.driver.switch_to_alert()
            alert_text = alert.text
            if self.accept_next_alert:
                alert.accept()
            else:
                alert.dismiss()
            return alert_text
        finally: self.accept_next_alert = True
    
    def tearDown(self):
        self.driver.quit()
        self.assertEqual([], self.verificationErrors)

if __name__ == "__main__":
    unittest.main()
